### PR TITLE
[Ready for review] use gfile to support remote directories

### LINKED ISSUE #2161: tensorboard logger should support remote directories
## 🚀 Feature
Tensorboard allows you to write to gc, s3, hdfs, etc by specifying paths with the right prefix  e.g. `logDir='hdfs://path/to/logs/` 

However the lightning logger breaks this. see [tensorboard.py#L99](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/tensorboard.py#L99)

### Motivation
Training often occurs on remote clusters which don't persist the local disk at the time the job ends. The local disk is also not accessible from outside tools so tensorboard can not access the results while training is in progress.

### Pitch
Replace all directory operations with some remote-aware tool. tensorboard itself provides a [gfile compatible handle .](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/compat/tensorflow_stub/io/gfile.py) There are other options as well. tensorboard itself supports these things natively so maybe we can just get around doing any local file operations and leverage tensorboard lib to write remotely.

### Alternatives
Some other option would be to write locally but add hooks to sync the logs to a remote storage.

### PR DESCRIPTION
**Pull request by f4hy**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Fixes #2161 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by f4hy** [comment]
Thanks @samj1912 fix the one missing call. I also cleaned up how this was being called and fixed up the tests. The tests were using the tmpdir pytest fixture which gives a `py.path.local` object where the contract in many places is type annotated `str` or `Optional[str]`. So I felt the best choice was to update the tests. Howerver it might be a better solution to just make the contracts some sort of path like object.

**Comment by f4hy** [comment]
Rebased since the requirements paths changed on master.

**Comment by f4hy** [comment]
I will leave this a draft PR in case a different solution is suggested for issue #2161

**Comment by f4hy** [comment]
Hmm. ok the tests here are now failing on just the 'minimal' builds where it passes on the lastest. A better approach here might be to wrap all the IO in something that can detect what is installed. If running on a minimum of dependencies only support local disk writes, and if more/later deps are installed support cloud file paths.

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -501,7 +501,7 @@ def test_benchmark_option(tmpdir):
 
     # fit model
     trainer = Trainer(
-        default_root_dir=tmpdir,
+        default_root_dir=str(tmpdir),
```
Comment (Borda) [inline tests/trainer/test_trainer.py:504]: rather so this casting in Trainer

Comment (f4hy) [reply tests/trainer/test_trainer.py:504]: Sounds good. will make that change.

Comment (f4hy) [reply tests/trainer/test_trainer.py:504]: The reason to do it here is that the trainer and other places _correctly_ type annotate that it should be a `str` or `Optional[str]` and it is hte tests which are wrong in passing a path object. @Borda I think if the tests are going to do this we should update the type annotations to accept a union of string and path. what do you think?

Comment (Borda) [reply tests/trainer/test_trainer.py]: yes, then pls update annotation :]

**Comment by williamFalcon** [comment]
@f4hy is this ready for 0.8.2 today?

**Comment by f4hy** [comment]
> @f4hy is this ready for 0.8.2 today?

Yep! Just removed some extra cruft. If you @williamFalcon think this is a good solution to this then I think its ready to go now. I was worried the tests are not passing on the 'minimal' builds, I tried to fix it but looks like its failing on master for minimal right now.

**Comment by Borda** [comment]
btw, how does it goes with #2175?

**Comment by f4hy** [comment]
> btw, how does it goes with #2175?

So s3 support is just one cloud hosting provider. My pr should support any of the backends that tensorboard currently supports which is `s3/hdfs/gcs/local` and maybe others have been added. So if #2175 is merged it wont solve the issue I have of #2161 . 

If the other PR does something this one does not we can add it here, but my understanding is that this is a superset of that.

**Comment by Borda** [comment]
@f4hy ho, how is it going, can we get this done in next release?

**Comment by f4hy** [comment]
@Borda Hey ya looks like I need to rebase an test again now. Let me get that done tomorrow. Hopefuly we can get this in soon.

**Comment by f4hy** [comment]
@Borda I have rebased master and updated the tests so now all pass except against windows. I am not quite sure how to test against windows and some tips would be welcome. I believe it is just due to line-ending issues. 

On a larger issue though, is this design acceptable? I know others have tried to solve similar problems. My issue is my training happens on a Kubernetes cluster where I only have remote file systems to store data so I want all logs/checkpoints/outputs to be written to remote/cloud storage. I think this is a reasonable way to approach it since tensorboard has this capability and is already a dependency but other approaches can work as well.

**Comment by f4hy** [comment]
@Borda Rebased again. Now passes all tests. Should be ready for review.

Let me know any ideas on how to improve this. I would like better ways of testing it. The concern I have is ensuring we don't later add some methods to write files that only work locally. So somehow testing this can actually work with remote files would be good, but not sure how to do that without something like docker compose to set that up. Ideas welcome.

**Comment by igorgad** [comment]
Hi, @f4hy. Was this already merged? I cannot see it on release 1.1.2. Did it got removed somehow?

**Comment by f4hy** [comment]
We went with fsspec instead. See the other PRs by me. Remote directories should be supported but with fsspec instead

**Pull request merged**